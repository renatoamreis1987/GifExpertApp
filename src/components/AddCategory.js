/*
AddCategory es el componente responsable de gestionar el buscador, 
en este caso va a obtener lo que ponemos en el buscador y guardarlo en el Hook useState, 
y lo vamos pasar al componente “GifExpertApp” en “setCategories((cats) => [inputValue, ...cats])”
*/


import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  // Hook useState
  const [inputValue, setinputValue] = useState("");

  const handleInputChange = (e) => {
    // Aquí vamos a añadir a inputValue el valor del input
    setinputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      // Aquí vamos a obtener las cats (desde un cb) y añadir el valor de input
      setCategories((cats) => [inputValue, ...cats]);
      setinputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p> { inputValue } </p>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange}
      ></input>
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
