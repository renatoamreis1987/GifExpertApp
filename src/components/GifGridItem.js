// GifGridItem es un simples componente que lo Ãºnico que va hacer es renderizar un div para cada GIF.

import React from "react";
import PropTypes from "prop-types";

export const GifGridItem = ({ title, url }) => {
  return (
    <div className="card animate__animated animate__fadeIn">
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
