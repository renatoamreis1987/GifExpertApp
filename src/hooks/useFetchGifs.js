/* ESTO ES UN CUSTOM HOOK 

En este CUSTOM HOOK lo que vamos hacer es llamar a useEffect, que es una helper Function, 
en la cual pasamos la categorÃ­a (la bÃºsqueda) y obtenemos una promesa con los resultados, 
y la almacenamos en nuestro useState y la enviamos como promesa con un return.  

*/

import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
  const [state, setstate] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(category).then((imgs) => {
      setstate({
        data: imgs,
        loading: false,
      });
    });
  }, [category]);

  return state; // {data: [], loading: true}
};
