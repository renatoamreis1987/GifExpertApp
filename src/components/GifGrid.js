/*
GifGrid es el componente responsable por renderizar el conjunto de GIFs de una categorÃ­a, 
va llamar a uno HOOKS -> useFetchGifs para obtener el listado, 
y una vez tenga ese listado lo va a renderizar llamando otro componente, GifGridItem. 
*/

import React from "react";
import PropTypes from 'prop-types';
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
  const { data: images, loading } = useFetchGifs(category);

  return (
    <>
      <h3 className='animate__animated animate__fadeIn'> {category}</h3>

      {loading && <p className='animate__animated animate__flash'>Loading</p>}

      <div className="card-grid">
        {images.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}

