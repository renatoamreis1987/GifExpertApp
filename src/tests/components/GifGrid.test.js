import React from "react";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock('../../hooks/useFetchGifs')

describe("Pruebas en <GifGrid />", () => {
  const category = "Batman";
  

  test("debe de mostrarse correctamente", () => {

    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    })
    let wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar items cuando se carga imÃ¡genes useFetchGifs', () => {
    
    const gifs = [{
      id: 'ABC',
      url: 'https://localhost/blablabla.jpg',
      title: 'Cualquier cosa'
    }]

    // Tenemos que simular lo que devuelve el Hook useFetchGifs
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    })
    let wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('p').exists()).toBe(false) // ya que tenemos data, que el loading no exista
    expect(wrapper.find('GifGridItem').length).toBe(gifs.length)
  })
  

});
