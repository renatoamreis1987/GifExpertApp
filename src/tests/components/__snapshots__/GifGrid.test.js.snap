// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> debe de mostrar items cuando se carga im√°genes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Batman
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/blablabla.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Batman
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
